server/src/main/java/org/elasticsearch/search/fetch/FetchPhase.java

```diff
@@ -524,4 +550,18 @@ private static void fillDocAndMetaFields(SearchContext context, FieldsVisitor fi
             }
         }
     }
+
+    /**
+     * Returns <code>true</code> if the provided <code>docs</code> are
+     * stored sequentially (Dn = Dn-1 + 1).
+     */
+    static boolean hasSequentialDocs(DocIdToIndex[] docs) {
+        for (int i = 1; i < docs.length; i++) {
+            assert docs[i].docId >= docs[i-1].docId : "doc ids out of order";
+            if (docs[i].docId - docs[i-1].docId > 1) {
+                return false;
+            }
+        }
+        return true;
```

Comment:
nit: since doc ids are sorted, we should be able to do something like `return docs.length > 0 && docs[docs.length-1].docId - docs[0].docId == docs.length - 1;`

Reply:
++, I pushed https://github.com/elastic/elasticsearch/pull/62509/commits/9d3f33fa1ad4877d55bfc1a13acd73111cfb6d46